jupyter/swan.py

```diff
@@ -0,0 +1,560 @@
+# --------------------------------
+
+
+CRIT = 'background:#a9341f; color: white;'
+WARN = 'background:#ffeda0'
+OK = 'background:#98cc70'
+NAN = 'background-color: #c0c0c0'
+
+
+def composite_qps_formatter(composite_values, normalized=False, show_qps=False):
+    if composite_values is None:
+        return 'N/A'
+
+    qps = composite_values[QPS_LABEL]
+    achieved_qps = composite_values[ACHIEVED_QPS_LABEL]
+
+    if any(map(pd.isnull, (achieved_qps, qps))):
+        return NAN
+
+    if normalized:
+        return '{:.0%}'.format(achieved_qps)
+    else:
+        return '{:,.0f}'.format(qps)
+
+
+def composite_qps_colors(composite_values, show_qps=False):
+    if pd.isnull(composite_values):
+        return NAN
+
+    achieved_qps = composite_values[ACHIEVED_QPS_LABEL]
+    if any(map(pd.isnull, (achieved_qps, ))):
+        return NAN
+
+    # format just according QPS value
+    if pd.isnull(achieved_qps):
+        return NAN
+    if achieved_qps > 0.95:
+        return OK
+    elif achieved_qps > 0.8:
+        return WARN
+    else:
+        return CRIT
+
+
+def composite_latency_formatter(composite_values, normalized=False):
+    if composite_values is None:
+        return 'N/A'
+
+    achieved_qps = composite_values[ACHIEVED_QPS_LABEL]
+    latency = composite_values[PERCENTILE99TH_LABEL]
+    achieved_latency = composite_values[ACHIEVED_LATENCY_LABEL]
+
+    if any(map(pd.isnull, (achieved_qps, latency, achieved_latency))):
+        return NAN
+
+    if achieved_qps < 0.9:
+        return 'FAIL'
+
+    if normalized:
+        if achieved_latency > 3.0:
+            return '> 300%'
+        else:
+            return '{:.0%}'.format(achieved_latency)
+    else:
+        return '{:.0f}'.format(latency)
+
+
+def composite_latency_colors(composite_values, slo):
+    if pd.isnull(composite_values):
+        return NAN
+
+    achieved_qps = composite_values[ACHIEVED_QPS_LABEL]
+    latency = composite_values[PERCENTILE99TH_LABEL]
+
+    if any(map(pd.isnull, (achieved_qps, latency))):
+        return NAN
+
+    # format just according QPS value
+    if pd.isnull(achieved_qps) or pd.isnull(latency):
+        return NAN
+
+    if achieved_qps < 0.9:
+        return NAN
+
+    if latency > 1.5 * slo:
+        return CRIT
+    elif latency > slo:
+        return WARN
+    else:
+        return OK
+
+
+class Experiment:
+    """ Base class for experiments that are based on load(QPS) and latency.
+
+    Responsible for loading data from cassandra, reshaping and preparing extra normalized columns
+    for latency/qps according target SLOs.
+
+    Extra normalized columns are then stored in one composite column for further processing.
+
+    Additionally provides a helper function to rename existing dataframe columns to new names and
+    function that allows to refere to new names using original names (_renamed).
+    """
+
+    def __init__(self, experiment_id, slo, cassandra_options=None, columns_to_rename=None, aggfunc=np.mean):
```

Comment:
@ppalucki - consider providing documentation for parameters (I would like to know what is the type of `columns_to_rename`).